src/coreclr/jit/morph.cpp

```diff
@@ -7069,6 +6990,168 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
     return result;
 }
 
+//------------------------------------------------------------------------
+// fgValidateIRForTailCall:
+//     Validate that the IR looks ok to perform a tailcall.
+//
+// Arguments:
+//     call - The call that we are dispatching as a tailcall.
+//
+// Notes:
+//   This function needs to handle somewhat complex IR that appears after
+//   tailcall candidates due to inlining.
+//
+void Compiler::fgValidateIRForTailCall(GenTreeCall* call)
+{
+#ifdef DEBUG
+    class TailCallIRValidatorVisitor final : public GenTreeVisitor<TailCallIRValidatorVisitor>
+    {
+        GenTreeCall* m_tailcall;
+        GenTree*     m_prevVal;
+
+    public:
+        enum
+        {
+            DoPostOrder       = true,
+            UseExecutionOrder = true,
+        };
+
+        TailCallIRValidatorVisitor(Compiler* comp, GenTreeCall* tailcall)
+            : GenTreeVisitor(comp), m_tailcall(tailcall), m_prevVal(nullptr)
+        {
+        }
+
+        fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+        {
+            GenTree* tree = *use;
+
+            // Wait until we get to the actual call...
+            if (m_prevVal == nullptr)
+            {
+                if (tree == m_tailcall)
+                {
+                    m_prevVal = m_tailcall;
+                }
+
+                return WALK_CONTINUE;
+            }
+
+            if (tree->OperIs(GT_RETURN))
+            {
+                assert((tree->TypeIs(TYP_VOID) || ValidateUse(tree->gtGetOp1())) &&
+                       "Expected return to be result of tailcall");
+                return WALK_ABORT;
+            }
+
+            // GT_NOP might appear due to assignments that end up as
+            // self-assignments, which get morphed to GT_NOP.
+            if (tree->OperIs(GT_NOP))
+            {
+            }
+            // No-op casts may appear due to normalization during inlining. Example:
+            //  ▌  RETURN    int
```

Comment:
Q: are we ok with using non-ASCII characters in source code?

Reply:
Made it into the ASCII format to be safe, although we do have some unicode chars in the VM it seems (the apostrophes): https://github.com/dotnet/runtime/blob/4719135a5fd6c9f22ce429ad8a211d696eae0426/src/coreclr/vm/codeman.cpp#L1331-L1334

Reply:
> Q: are we ok with using non-ASCII characters in source code?

Yup, it is safe. We have [`source-charset`](https://docs.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170) set to `utf-8` globally:
https://github.com/dotnet/runtime/blob/842b6a87aaa6b6f25555541f541d5ee202d3972a/eng/native/configurecompiler.cmake#L648 (and Unix-y toolchains have UTF-8 by default).